src/main/java/com/nextcloud/android/sso/api/AidlNetworkRequest.java

```diff
@@ -218,10 +247,128 @@ public void onThreadFinished(Thread thread) {
+     * DO NOT CALL THIS METHOD DIRECTLY - use @link(performNetworkRequest) instead
+     *
+     * @param request
+     * @return
+     * @throws IOException
+     */
+    private ParcelFileDescriptor performNewAidlNetworkRequest(NextcloudRequest request,
+                                                              InputStream requestBodyInputStream)
+            throws IOException, RemoteException, NextcloudApiNotRespondingException {
+
+        // Check if we are on the main thread
+        if (Looper.myLooper() == Looper.getMainLooper()) {
+            throw new NetworkOnMainThreadException();
+        }
+
+        // Wait for api to be initialized
+        waitForApi();
+
+        // Log.d(TAG, request.url);
+        request.setAccountName(getAccountName());
+        request.setToken(getAccountToken());
+
+        ByteArrayOutputStream baos = new ByteArrayOutputStream();
+        ObjectOutputStream oos = new ObjectOutputStream(baos);
+        oos.writeObject(request);
+        oos.close();
+        baos.close();
+        InputStream is = new ByteArrayInputStream(baos.toByteArray());
+
+        ParcelFileDescriptor input = ParcelFileDescriptorUtil.pipeFrom(is,
+                new IThreadListener() {
+                    @Override
+                    public void onThreadFinished(Thread thread) {
+                        Log.d(TAG, "copy data from service finished");
+                    }
+                });
+
+        ParcelFileDescriptor requestBodyParcelFileDescriptor = null;
+        if (requestBodyInputStream != null) {
+            requestBodyParcelFileDescriptor = ParcelFileDescriptorUtil.pipeFrom(requestBodyInputStream,
+                    new IThreadListener() {
+                        @Override
+                        public void onThreadFinished(Thread thread) {
+                            Log.d(TAG, "copy data from service finished");
+                        }
+                    });
+        }
+
+        ParcelFileDescriptor output;
+        if (requestBodyParcelFileDescriptor != null) {
+            output = mService.performNewNextcloudRequestAndBodyStream(input, requestBodyParcelFileDescriptor);
+        } else {
+            output = mService.performNewNextcloudRequest(input);
+        }
+
+        return output;
+    }
 
     private static <T> T deserializeObject(InputStream is) throws IOException, ClassNotFoundException {
         ObjectInputStream ois = new ObjectInputStream(is);
         T result = (T) ois.readObject();
         return result;
     }
+
+    private ExceptionResponse deserializeNewObject(InputStream is) throws IOException, ClassNotFoundException {
+        ObjectInputStream ois = new ObjectInputStream(is);
+        ArrayList<PlainHeader> headerList = new ArrayList<>();
+        Exception exception = (Exception) ois.readObject();
+
+        if (exception == null) {
+            String headers = (String) ois.readObject();
+            ArrayList list = new Gson().fromJson(headers, ArrayList.class);
```

Comment:
In terms of performance it might make sense to create one instance of the Gson object and reuse it in all requests? If I remember correctly creating a new Gson parser is a quite costly operation. And I think we can even parse those values already as string by using:

```java
list = gson.fromJson(headers, new TypeToken<List<String>>(){}.getType());
```

https://stackoverflow.com/a/12384156/11405479

Reply:
Those are name value pairs, not plain strings, so this is not possible